README
NAME
    Dancer2::Plugin::Deferred - Defer messages or data across redirections

VERSION
    version 0.005

SYNOPSIS
      use Dancer2::Plugin::Deferred;

      get '/defer' => sub {
        deferred error => "Klaatu barada nikto";
        redirect '/later';
      }

      get '/later' => sub {
        template 'later';
      }

      # in template 'later.tt'
      <% IF deferred.error %>
      <div class="error"><% deferred.error %></div>
      <% END %>

DESCRIPTION
    This Dancer2 plugin provides a method for deferring a one-time message
    across a redirect. It is similar to "flash" messages, but without the
    race conditions that can result from multiple tabs in a browser or from
    AJAX requests. It is similar in design to
    Catalyst::Plugin::StatusMessage, but adapted for Dancer2.

    It works by creating a unique message ID within the session that holds
    deferred data. The message ID is automatically added as a query
    parameter to redirection requests. It's sort of like a session within a
    session, but tied to a request rather than global to the browser. (It
    will even chain across multiple redirects.)

    When a template is rendered, a pre-template hook retrieves the data and
    deletes it from the session. Alternatively, the data can be retrieved
    manually (which will also automatically delete the data.)

    Alternatively, the message ID parameters can be retrieved and used to
    construct a hyperlink for a message to be retrieved later. In this case,
    the message is preserved past the template hook. (The template should be
    sure not to render the message if not desired.)

USAGE
  deferred
      deferred $key => $value;
      $value = deferred $key; # also deletes $key

    This function works just like "var" or "session", except that it lasts
    only for the current request and across any redirects. Data is deleted
    if accessed. If a key is set to an undefined value, the key is deleted
    from the deferred data hash.

  all_deferred
      template 'index', { deferred => all_deferred };

    This function returns all the deferred data as a hash reference and
    deletes the stored data. This is called automatically in the
    "before_template_render" hook, but is available if someone wants to have
    manual control.

  deferred_param
      template 'index' => { link => uri_for( '/other', { deferred_param } ) };

    This function returns the parameter key and value used to propagate the
    message to another request. Using this function toggles the
    "var_keep_key" variable to true to ensure the message remains to be
    retrieved by the link.

CONFIGURATION
    *   "var_key: dpdid" -- this is the key in the "var" hash containing the
        message ID

    *   "var_keep_key: dpd_keep" -- if this key in "var" is true, retrieving
        values will not be destructive

    *   "params_key: dpdid" -- this is the key in the "params" hash
        containing the message ID

    *   "session_key_prefix": dpd_> -- the message ID is appended to this
        prefix and used to store deferred data in the session

    *   "template_key: deferred" -- this is the key to deferred data passed
        to the template

SEE ALSO
    *   Dancer2

    *   Dancer::Plugin::FlashMessage

    *   Dancer::Plugin::FlashNote

    *   Catalyst::Plugin::StatusMessage

ACKNOWLEDGMENTS
    Thank you to mst for explaining why Catalyst::Plugin::StatusMessages
    does what it does and putting up with my dumb ideas along the way.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://github.com/dagolden/dancer2-plugin-deferred/issues>. You will
    be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/dagolden/dancer2-plugin-deferred>

      git clone git://github.com/dagolden/dancer2-plugin-deferred.git

AUTHOR
    David Golden <dagolden@cpan.org>

CONTRIBUTOR
    Damien Krotkine <dams@zarb.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by David Golden.

    This is free software, licensed under:

      The Apache License, Version 2.0, January 2004



Changes
Revision history for Dancer2-Plugin-Deferred

0.005     2013-07-20 16:02:39 America/New_York

  [CHANGED]

  - Updated internal workarounds to use new features in Dancer2. Bumped
    Dancer2 prereq to 0.05.

0.004     2013-03-06 20:48:23 America/New_York

  [CHANGED]

  - Forked into Dancer2::Plugin::Deferred for proper Dancer2 support

***** Previous history applies to Dancer::Plugin::Deferred *****

0.003     2012-12-26 11:31:06 America/New_York

  [FIXED]

  - Tests fixed for changes in Dancer 2 alpha

0.002     2012-12-20 12:31:41 America/New_York

  [ADDED]

  - Added abillity to preserve the message and attach message
    parameters manually to an arbitrary URI.

0.001     2012-12-19 21:56:38 America/New_York

  - First release

# vim: ts=2 sts=2 sw=2 et:


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Defer messages or data across redirections",
  "AUTHOR" => "David Golden <dagolden\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Dancer2-Plugin-Deferred",
  "EXE_FILES" => [],
  "LICENSE" => "apache",
  "NAME" => "Dancer2::Plugin::Deferred",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Dancer2::Plugin" => "0.05",
    "URI" => 0,
    "URI::QueryParam" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer2" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "LWP::UserAgent" => 0,
    "List::Util" => 0,
    "Test::More" => "0.96",
    "Test::TCP" => 0
  },
  "VERSION" => "0.005",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Dancer2/Plugin/Deferred.pm
use 5.008001;
use strict;
use warnings;

package Dancer2::Plugin::Deferred;
# ABSTRACT: Defer messages or data across redirections
our $VERSION = '0.005'; # VERSION

use Carp qw/croak/;
use URI;
use URI::QueryParam;

use Dancer2::Plugin 0.05;

my $conf;

register 'deferred' => sub {
    my ( $dsl, $key, $value ) = plugin_args(@_);
    $conf ||= _get_conf();
    my $app     = $dsl->app;
    my $context = $app->context;
    my $id      = _get_id($dsl);

    # message data is flat "dpd_$id" to avoid race condition with
    # another session
    my $data = $app->session( $conf->{session_key_prefix} . $id ) || {};

    # set value or destructively retrieve it
    if ( defined $value ) {
        $data->{$key} = $value;
    }
    else {
        $value =
          $context->var( $conf->{var_keep_key} ) ? $data->{$key} : delete $data->{$key};
    }

    # store remaining data or clear it if no deferred messages are left
    if ( keys %$data ) {
        $app->session( $conf->{session_key_prefix} . $id => $data );
        $context->var( $conf->{var_key} => $id );
    }
    else {
        $app->session->delete( $conf->{session_key_prefix} . $id );
        $context->var( $conf->{var_key} => undef );
    }

    return $value;
};

# destructively return all keys
register 'all_deferred' => \&_get_all_deferred;

sub _get_all_deferred {
    my $dsl     = shift;
    my $app     = $dsl->app;
    my $context = $app->context;

    my $id = _get_id($dsl);
    my $data = $app->session( $conf->{session_key_prefix} . $id ) || {};
    unless ( $context->var( $conf->{var_keep_key} ) ) {
        $app->session->delete( $conf->{session_key_prefix} . $id );
        $context->var( $conf->{var_key}, undef );
    }
    return $data;
}

register 'deferred_param' => \&_get_deferred_param;

sub _get_deferred_param {
    my $dsl = shift;
    $conf ||= _get_conf();
    $dsl->app->context->var( $conf->{var_keep_key} => 1 );
    return ( $conf->{params_key} => $dsl->app->context->var( $conf->{var_key} ) );
}

# not crypto strong, but will be stored in session, which should be
sub _get_id {
    my $dsl = shift;
    $conf ||= _get_conf();
    return $dsl->app->context->var( $conf->{var_key} )
      || sprintf( "%08d", int( rand(100_000_000) ) );
}

sub _get_conf {
    return {
        var_key            => 'dpdid',
        var_keep_key       => 'dpd_keep',
        params_key         => 'dpdid',
        session_key_prefix => 'dpd_',
        template_key       => 'deferred',
        %{ plugin_setting() },
    };
}

on_plugin_import {
    my $dsl = shift;

    $dsl->app->add_hook(
        Dancer2::Core::Hook->new(
            name => 'before_template',
            code => sub {
                my $data = shift;
                $conf ||= _get_conf();
                $data->{ $conf->{template_key} } = _get_all_deferred($dsl);
            }
        )
    );

    $dsl->app->add_hook(
        Dancer2::Core::Hook->new(
            name => 'before',
            code => sub {
                $conf ||= _get_conf();
                my $id = $dsl->context->request->params->{ $conf->{params_key} };
                $dsl->app->context->var( $conf->{var_key} => $id )
                  if $id;
            }
        )
    );

    $dsl->app->add_hook(
        Dancer2::Core::Hook->new(
            name => 'after',
            code => sub {
                my $response = shift;
                $conf ||= _get_conf();
                if ( $dsl->app->context->var( $conf->{var_key} ) && $response->status =~ /^3/ ) {
                    my $u = URI->new( $response->header("Location") );
                    $u->query_param( _get_deferred_param($dsl) );
                    $response->header( "Location" => $u );
                }
            }
        )
    );
};

register_plugin;

1;


# vim: ts=4 sts=4 sw=4 et:

__END__

=pod

=encoding utf-8

=head1 NAME

Dancer2::Plugin::Deferred - Defer messages or data across redirections

=head1 VERSION

version 0.005

=head1 SYNOPSIS

  use Dancer2::Plugin::Deferred;

  get '/defer' => sub {
    deferred error => "Klaatu barada nikto";
    redirect '/later';
  }

  get '/later' => sub {
    template 'later';
  }

  # in template 'later.tt'
  <% IF deferred.error %>
  <div class="error"><% deferred.error %></div>
  <% END %>

=head1 DESCRIPTION

This L<Dancer2> plugin provides a method for deferring a one-time message across
a redirect.  It is similar to "flash" messages, but without the race conditions
that can result from multiple tabs in a browser or from AJAX requests.  It is
similar in design to L<Catalyst::Plugin::StatusMessage>, but adapted for Dancer2.

It works by creating a unique message ID within the session that holds deferred
data.  The message ID is automatically added as a query parameter to redirection
requests.  It's sort of like a session within a session, but tied to a request
rather than global to the browser.  (It will even chain across multiple
redirects.)

When a template is rendered, a pre-template hook retrieves the data and
deletes it from the session.  Alternatively, the data can be retrieved manually
(which will also automatically delete the data.)

Alternatively, the message ID parameters can be retrieved and used to
construct a hyperlink for a message to be retrieved later.  In this case,
the message is preserved past the template hook.  (The template should be
sure not to render the message if not desired.)

=for Pod::Coverage method_names_here

=head1 USAGE

=head2 deferred

  deferred $key => $value;
  $value = deferred $key; # also deletes $key

This function works just like C<var> or C<session>, except that it lasts only
for the current request and across any redirects.  Data is deleted if accessed.
If a key is set to an undefined value, the key is deleted from the deferred
data hash.

=head2 all_deferred

  template 'index', { deferred => all_deferred };

This function returns all the deferred data as a hash reference and deletes
the stored data.  This is called automatically in the C<before_template_render>
hook, but is available if someone wants to have manual control.

=head2 deferred_param

  template 'index' => { link => uri_for( '/other', { deferred_param } ) };

This function returns the parameter key and value used to propagate the
message to another request.  Using this function toggles the C<var_keep_key>
variable to true to ensure the message remains to be retrieved by the link.

=head1 CONFIGURATION

=over 4

=item *

C<var_key: dpdid> -- this is the key in the C<var> hash containing the message ID

=item *

C<var_keep_key: dpd_keep> -- if this key in C<var> is true, retrieving values will not be destructive

=item *

C<params_key: dpdid> -- this is the key in the C<params> hash containing the message ID

=item *

C<session_key_prefix>: dpd_> -- the message ID is appended to this prefix and used to store deferred data in the session

=item *

C<template_key: deferred> -- this is the key to deferred data passed to the template

=back

=head1 SEE ALSO

=over 4

=item *

L<Dancer2>

=item *

L<Dancer::Plugin::FlashMessage>

=item *

L<Dancer::Plugin::FlashNote>

=item *

L<Catalyst::Plugin::StatusMessage>

=back

=head1 ACKNOWLEDGMENTS

Thank you to mst for explaining why L<Catalyst::Plugin::StatusMessages> does
what it does and putting up with my dumb ideas along the way.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/dagolden/dancer2-plugin-deferred/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/dagolden/dancer2-plugin-deferred>

  git clone git://github.com/dagolden/dancer2-plugin-deferred.git

=head1 AUTHOR

David Golden <dagolden@cpan.org>

=head1 CONTRIBUTOR

Damien Krotkine <dams@zarb.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by David Golden.

This is free software, licensed under:

  The Apache License, Version 2.0, January 2004

=cut
